File: utils/dataFactory.ts
import { faker } from "@faker-js/faker";

export type UserPayload = {
  nome: string;
  email: string;
  password: string;
  administrador: "true" | "false";
};

export type ProductPayload = {
  nome: string;
  preco: number;
  descricao: string;
  quantidade: number;
};

export const createUserPayload = (overrides: Partial<UserPayload> = {}): UserPayload => ({
  nome: faker.person.fullName(),
  email: faker.internet.email().toLowerCase(),
  password: faker.internet.password({ length: 10 }),
  administrador: "true",
  ...overrides
});

export const createProductPayload = (
  overrides: Partial<ProductPayload> = {}
): ProductPayload => ({
  nome: `${faker.commerce.productName()} ${Date.now()}`,
  preco: faker.number.int({ min: 50, max: 500 }),
  descricao: faker.commerce.productDescription(),
  quantidade: faker.number.int({ min: 1, max: 50 }),
  ...overrides
});

File: src/test/java/com/serverest/utils/DataFactory.java
package com.serverest.utils;

import com.github.javafaker.Faker;
import com.serverest.model.Product;
import com.serverest.model.User;

import java.util.Locale;
import java.util.UUID;

public final class DataFactory {
    private static final Faker FAKER = new Faker(new Locale("pt", "BR"));

    private DataFactory() {
    }

    public static User createAdminUser() {
        return User.builder()
                .nome(FAKER.name().fullName())
                .email(uniqueEmail())
                .password(FAKER.internet().password(10, 16))
                .administrador("true")
                .build();
    }

    public static User createDefaultUser() {
        return User.builder()
                .nome(FAKER.name().fullName())
                .email(uniqueEmail())
                .password(FAKER.internet().password(10, 16))
                .administrador("false")
                .build();
    }

    public static Product createProduct() {
        return Product.builder()
                .nome("test_" + RunContext.runId() + "_" + FAKER.commerce().productName() + " " + UUID.randomUUID().toString().substring(0, 8))
                .preco(FAKER.number().numberBetween(50, 500))
                .descricao(FAKER.commerce().material())
                .quantidade(FAKER.number().numberBetween(1, 50))
                .build();
    }

    private static String uniqueEmail() {
        String baseLocal = FAKER.internet().emailAddress().split("@")[0];
        return RunContext.emailWithRunId(baseLocal);
    }
}
